File: agent/marketmaker.go
package agent

import (
	"math"

	"economy/company"
	"economy/market"
)

// MarketMaker provides continuous two-sided liquidity, deriving its
// quotes from an inventory-based model (Avellaneda-Stoikov), adapted
// for a discrete-tick simulation rather than continuous time.
//
// Three things distinguish this from a naive "quote around the mid"
// implementation, and all three are necessary for realistic behavior,
// not independent nice-to-haves:
//
//  1. Fair value is the MM's own EMA of past mid-prices, not the
//     current mid itself. This gives it genuine informational lag —
//     during a fast one-directional move, its quotes trail the market,
//     so it keeps buying while price falls (or selling while it
//     rises), accumulating inventory it doesn't want. Real market
//     makers get run over this way; a MM with no independent fair
//     value estimate cannot exhibit that failure mode at all.
//
//  2. The reservation price (the true center of its quotes) is shifted
//     away from fair value by its current inventory, scaled by both
//     risk aversion AND volatility — holding a large position is more
//     dangerous when volatility is high, so the skew has to respond to
//     both, not inventory alone.
//
//  3. Spread width itself grows with volatility (and with risk
//     aversion), which is what naturally throttles the MM's fill rate
//     and presence as conditions worsen — continuous widening rather
//     than a binary on/off pull-out, matching how real liquidity
//     provision degrades under stress.
type MarketMaker struct {
	id      string
	account *market.Account

	// fairValue is the MM's own running estimate of fair value,
	// updated once per tick from the observed mid-price. Distinct
	// from MarketState.Mid on purpose — see type doc.
	fairValue *EMA

	// company is an optional reference to the covered enterprise fundamentals.
	// When present, fairValue is anchored to genuine business performance (ReportedValue).
	company *company.Company

	// LadderLevels specifies how many price depth levels to quote on each side (defaults to 1).
	LadderLevels int

	// RiskAversion (γ) controls how strongly inventory and volatility
	// widen the spread and shift the reservation price. Higher =
	// more risk-averse: wider spreads, more aggressive inventory
	// unwinding pressure.
	RiskAversion float64

	// BaseHalfSpread is the minimum half-spread quoted even at zero
	// volatility and zero inventory — covers a MM's baseline cost of
	// operating (adverse selection risk that exists even in calm
	// markets).
	BaseHalfSpread float64

	// SpreadVolCoefficient (k) scales how much variance contributes
	// to spread widening: halfSpread = BaseHalfSpread + k*variance.
	SpreadVolCoefficient float64

	// MaxInventory is the position size (in units of the instrument,
	// signed) beyond which the MM stops quoting on the side that
	// would increase exposure further, but will still quote (often
	// more aggressively) on the side that reduces it.
	MaxInventory float64

	// QuoteSize is the size posted at best bid/ask each tick. Kept
	// fixed for now — a natural later refinement is scaling this down
	// as inventory approaches MaxInventory.
	QuoteSize float64
}

// NewMarketMaker constructs a MarketMaker with an initial cash
// account. maxLeverage/maintenanceMarginRatio follow the same meaning
// as any other market.Account — a MM typically runs high leverage
// relative to a directional trader, since its risk is managed via
// inventory limits and spread, not position sizing alone.
func NewMarketMaker(id string, startingCash, maxLeverage, maintenanceMarginRatio float64) *MarketMaker {
	return &MarketMaker{
		id:                   id,
		account:              market.NewAccount(id, startingCash, maxLeverage, maintenanceMarginRatio),
		fairValue:            NewEMA(0.1),
		LadderLevels:         1,
		RiskAversion:         0.1,
		BaseHalfSpread:       0.02,
		SpreadVolCoefficient: 50.0,
		MaxInventory:         1000,
		QuoteSize:            100,
	}
}

func (m *MarketMaker) ID() string               { return m.id }
func (m *MarketMaker) Account() *market.Account { return m.account }

// SetCompany associates a company with this market maker so quote ladders track fundamental enterprise performance.
func (m *MarketMaker) SetCompany(c *company.Company) {
	m.company = c
}

// reservationPrice computes the inventory- and volatility-adjusted
// center of the MM's quotes: r = fairValue - inventory*γ*σ².
// A positive inventory (net long) pulls the reservation price down —
// the MM wants to encourage selling and discourage further buying, so
// its whole quote ladder shifts down. A negative inventory (net
// short) pushes it up, for the symmetric reason.
func (m *MarketMaker) reservationPrice(inventory float64, variance float64) float64 {
	effectiveVar := math.Max(variance, 0.0002)
	return m.fairValue.Value() - inventory*m.RiskAversion*effectiveVar
}

// halfSpread computes quote half-width: widens with both configured
// risk aversion and current variance, so degraded conditions
// (high RecentVolatility) produce wider quotes and therefore lower
// effective participation, without a hard on/off switch.
func (m *MarketMaker) halfSpread(variance float64) float64 {
	return m.BaseHalfSpread + m.SpreadVolCoefficient*m.RiskAversion*variance
}

// NextOrders implements market.OrderSource. A market maker with a
// valid fair-value estimate quotes both sides in the same tick,
// subject to inventory limits: it will not add to a side that would
// push its position further past MaxInventory in the direction it is
// already extended, but continues to quote the opposite (risk-
// reducing) side even when the risk-increasing side is blocked. This
// mirrors a real desk's behavior under an inventory limit breach —
// it doesn't stop trading altogether, it stops trading in the
// direction that makes things worse.
func (m *MarketMaker) NextOrders(state market.MarketState) []*market.Order {
	if state.HasMid {
		m.fairValue.Update(state.Mid)
	}

	var center float64
	if m.company != nil {
		fund := m.company.ReportedValue
		if fund <= 0 {
			fund = m.company.TrueValue
		}
		if fund > 0 {
			if m.fairValue.Initialized() {
				center = 0.70*fund + 0.30*m.fairValue.Value()
			} else {
				center = fund
			}
		}
	}
	if center <= 0 {
		if !m.fairValue.Initialized() {
			return nil // no fair value estimate yet, nothing to quote around
		}
		center = m.fairValue.Value()
	}

	variance := Variance(state.RecentVolatility)
	inventory := NetInventory(m.account, state.Symbol)

	effectiveVar := math.Max(variance, 0.0002)
	skew := inventory * m.RiskAversion * effectiveVar
	maxSkew := center * 0.04
	if skew > maxSkew {
		skew = maxSkew
	} else if skew < -maxSkew {
		skew = -maxSkew
	}
	r := center - skew

	maxHalfSpread := math.Max(0.10, center*0.02)
	halfSpread := math.Min(m.halfSpread(variance), maxHalfSpread)

	// Dynamic inventory management:
	// Rather than completely pulling one side (which collapses market mid-price and halts trading),
	// dynamically scale quote sizes and skew spreads so the MM continuously maintains a 2-sided book.
	buySize := m.QuoteSize
	sellSize := m.QuoteSize

	if inventory > m.MaxInventory*0.75 {
		// Long inventory high: shrink buy quote to minimum, expand sell quote to shed shares
		excess := (inventory - m.MaxInventory*0.75) / (m.MaxInventory * 0.25)
		buySize = math.Max(1.0, math.Floor(m.QuoteSize*(1.0-0.9*math.Min(1.0, excess))))
		sellSize = math.Floor(m.QuoteSize * (1.0 + 0.5*math.Min(2.0, excess)))
	} else if inventory < -m.MaxInventory*0.75 {
		// Short inventory high: shrink sell quote, expand buy quote to cover shares
		excess := (-inventory - m.MaxInventory*0.75) / (m.MaxInventory * 0.25)
		sellSize = math.Max(1.0, math.Floor(m.QuoteSize*(1.0-0.9*math.Min(1.0, excess))))
		buySize = math.Floor(m.QuoteSize * (1.0 + 0.5*math.Min(2.0, excess)))
	}

	canBuy := inventory < m.MaxInventory
	canSell := inventory > -m.MaxInventory

	if m.LadderLevels <= 1 {
		bidPrice := math.Round((r-halfSpread)*100) / 100
		askPrice := math.Round((r+halfSpread)*100) / 100
		if bidPrice < 0.01 {
			bidPrice = 0.01
		}
		if askPrice <= bidPrice {
			askPrice = bidPrice + 0.01
		}

		var orders []*market.Order
		if canBuy {
			orders = append(orders, &market.Order{
				AgentID:  m.id,
				Side:     market.Buy,
				Price:    bidPrice,
				Quantity: buySize,
				IsMarket: false,
			})
		}
		if canSell {
			orders = append(orders, &market.Order{
				AgentID:  m.id,
				Side:     market.Sell,
				Price:    askPrice,
				Quantity: sellSize,
				IsMarket: false,
			})
		}
		return orders
	}

	// Multi-level depth ladder: provides deep, mobile liquidity that travels with the market
	ladderSteps := []struct {
		offsetMult float64
		sizeMult   float64
	}{
		{1.0, 1.0},
		{2.5, 1.5},
		{5.0, 2.5},
		{9.0, 4.0},
		{15.0, 6.0},
	}
	if m.LadderLevels < len(ladderSteps) {
		ladderSteps = ladderSteps[:m.LadderLevels]
	}

	var orders []*market.Order
	for _, step := range ladderSteps {
		stepOffset := halfSpread * step.offsetMult
		bp := math.Round((r-stepOffset)*100) / 100
		ap := math.Round((r+stepOffset)*100) / 100
		if bp < 0.01 {
			bp = 0.01
		}
		if ap <= bp {
			ap = bp + 0.01
		}

		if canBuy {
			orders = append(orders, &market.Order{
				AgentID:  m.id,
				Side:     market.Buy,
				Price:    bp,
				Quantity: math.Round(buySize * step.sizeMult),
				IsMarket: false,
			})
		}
		if canSell {
			orders = append(orders, &market.Order{
				AgentID:  m.id,
				Side:     market.Sell,
				Price:    ap,
				Quantity: math.Round(sellSize * step.sizeMult),
				IsMarket: false,
			})
		}
	}
	return orders
}

// EffectiveSpread is exposed for logging/analysis: lets the sim loop
// or a research notebook record how wide this MM was actually quoting
// at a given tick's volatility, without duplicating the formula.
func (m *MarketMaker) EffectiveSpread(recentVolatility float64) float64 {
	return 2 * m.halfSpread(Variance(recentVolatility))
}

// EffectiveReservationPrice mirrors EffectiveSpread for the
// reservation price, using current inventory for whichever symbol is
// passed in.
func (m *MarketMaker) EffectiveReservationPrice(symbol string, recentVolatility float64) float64 {
	return m.reservationPrice(NetInventory(m.account, symbol), Variance(recentVolatility))
}

File: src/agent/marketmaker.rs
use std::sync::Arc;
use parking_lot::RwLock;

use crate::company::Company;
use crate::market::{Account, MarketState, Order, OrderSource, Side};
use super::agent::{net_inventory, variance, EMA};

pub struct MarketMaker {
    pub id: String,
    pub account: Arc<RwLock<Account>>,
    pub fair_value: EMA,
    pub company: Option<Company>,
    pub ladder_levels: usize,
    pub risk_aversion: f64,
    pub base_half_spread: f64,
    pub spread_vol_coefficient: f64,
    pub max_inventory: f64,
    pub quote_size: f64,
}

impl MarketMaker {
    pub fn new(
        id: impl Into<String>,
        starting_cash: f64,
        max_leverage: f64,
        maintenance_margin_ratio: f64,
    ) -> Self {
        let id_str = id.into();
        Self {
            account: Arc::new(RwLock::new(Account::new(
                id_str.clone(),
                starting_cash,
                max_leverage,
                maintenance_margin_ratio,
            ))),
            id: id_str,
            fair_value: EMA::new(0.1),
            company: None,
            ladder_levels: 1,
            risk_aversion: 0.1,
            base_half_spread: 0.02,
            spread_vol_coefficient: 50.0,
            max_inventory: 1000.0,
            quote_size: 100.0,
        }
    }

    pub fn set_company(&mut self, company: Company) {
        self.company = Some(company);
    }

    pub fn reservation_price(&self, inventory: f64, variance_val: f64) -> f64 {
        let effective_var = variance_val.max(0.0002);
        self.fair_value.value() - inventory * self.risk_aversion * effective_var
    }

    pub fn half_spread(&self, variance_val: f64) -> f64 {
        self.base_half_spread + self.spread_vol_coefficient * self.risk_aversion * variance_val
    }

    pub fn effective_spread(&self, recent_volatility: f64) -> f64 {
        2.0 * self.half_spread(variance(recent_volatility))
    }

    pub fn effective_reservation_price(&self, symbol: &str, recent_volatility: f64) -> f64 {
        let acct = self.account.read();
        self.reservation_price(net_inventory(&acct, symbol), variance(recent_volatility))
    }
}

impl OrderSource for MarketMaker {
    fn id(&self) -> &str {
        &self.id
    }

    fn next_orders(&mut self, state: &MarketState) -> Vec<Order> {
        if let Some(mid) = state.mid {
            self.fair_value.update(mid);
        }

        let mut center = 0.0;
        let fund = if state.fundamental_value > 0.0 {
            state.fundamental_value
        } else if let Some(ref co) = self.company {
            if co.reported_value > 0.0 { co.reported_value } else { co.true_value }
        } else {
            0.0
        };

        if fund > 0.0 {
            center = if self.fair_value.initialized() {
                0.70 * fund + 0.30 * self.fair_value.value()
            } else {
                fund
            };
        }
        if center <= 0.0 {
            if !self.fair_value.initialized() {
                return Vec::new();
            }
            center = self.fair_value.value();
        }

        let var_val = variance(state.recent_volatility);
        let inventory = {
            let acct = self.account.read();
            net_inventory(&acct, &state.symbol)
        };

        let effective_var = var_val.max(0.0002);
        let mut skew = inventory * self.risk_aversion * effective_var;
        let max_skew = center * 0.04;
        skew = skew.clamp(-max_skew, max_skew);

        let r = center - skew;
        let max_half_spread = 0.10f64.max(center * 0.02);
        let half_spread = self.half_spread(var_val).min(max_half_spread);

        let mut buy_size = self.quote_size;
        let mut sell_size = self.quote_size;

        if inventory > self.max_inventory * 0.75 {
            let excess = (inventory - self.max_inventory * 0.75) / (self.max_inventory * 0.25);
            buy_size = 1.0f64.max((self.quote_size * (1.0 - 0.9 * excess.min(1.0))).floor());
            sell_size = (self.quote_size * (1.0 + 0.5 * excess.min(2.0))).floor();
        } else if inventory < -self.max_inventory * 0.75 {
            let excess = (-inventory - self.max_inventory * 0.75) / (self.max_inventory * 0.25);
            sell_size = 1.0f64.max((self.quote_size * (1.0 - 0.9 * excess.min(1.0))).floor());
            buy_size = (self.quote_size * (1.0 + 0.5 * excess.min(2.0))).floor();
        }

        let can_buy = inventory < self.max_inventory;
        let can_sell = inventory > -self.max_inventory;

        if self.ladder_levels <= 1 {
            let mut bid_price = ((r - half_spread) * 100.0).round() / 100.0;
            let mut ask_price = ((r + half_spread) * 100.0).round() / 100.0;
            if bid_price < 0.01 {
                bid_price = 0.01;
            }
            if ask_price <= bid_price {
                ask_price = bid_price + 0.01;
            }

            let mut orders = Vec::new();
            if can_buy {
                orders.push(Order {
                    id: 0,
                    agent_id: self.id.clone(),
                    side: Side::Buy,
                    price: bid_price,
                    quantity: buy_size,
                    is_market: false,
                });
            }
            if can_sell {
                orders.push(Order {
                    id: 0,
                    agent_id: self.id.clone(),
                    side: Side::Sell,
                    price: ask_price,
                    quantity: sell_size,
                    is_market: false,
                });
            }
            return orders;
        }

        let ladder_steps = [
            (1.0, 1.0),
            (2.5, 1.5),
            (5.0, 2.5),
            (9.0, 4.0),
            (15.0, 6.0),
        ];
        let steps_to_use = &ladder_steps[..self.ladder_levels.min(ladder_steps.len())];

        let mut orders = Vec::new();
        for (offset_mult, size_mult) in steps_to_use {
            let step_offset = half_spread * offset_mult;
            let mut bp = ((r - step_offset) * 100.0).round() / 100.0;
            let mut ap = ((r + step_offset) * 100.0).round() / 100.0;
            if bp < 0.01 {
                bp = 0.01;
            }
            if ap <= bp {
                ap = bp + 0.01;
            }

            if can_buy {
                orders.push(Order {
                    id: 0,
                    agent_id: self.id.clone(),
                    side: Side::Buy,
                    price: bp,
                    quantity: (buy_size * size_mult).round(),
                    is_market: false,
                });
            }
            if can_sell {
                orders.push(Order {
                    id: 0,
                    agent_id: self.id.clone(),
                    side: Side::Sell,
                    price: ap,
                    quantity: (sell_size * size_mult).round(),
                    is_market: false,
                });
            }
        }

        orders
    }
}
